lib/SILGen/SILGenLValue.cpp

```diff
@@ -1344,6 +1351,86 @@ namespace {
       assert(getAccessorDecl()->isSetter());
       SILDeclRef setter = Accessor;
 
+      if (hasPropertyDelegate() && IsOnSelfParameter) {
+        // This is delegated property. Instead of emitting a setter, emit an
+        // assign_by_delegate with the allocating initializer funcction and the
+        // setter function as arguments. DefiniteInitializtion will then decide
+        // between the two functions, depending if it's an initialization or a
+        // re-assignment.
+        //
+        VarDecl *field = dyn_cast<VarDecl>(Storage);
+        VarDecl *backingVar = field->getPropertyDelegateBackingVar();
+        assert(backingVar);
+        CanType ValType = backingVar->getType()->getCanonicalType();
+        SILType varStorageType =
+          SGF.SGM.Types.getSubstitutedStorageType(backingVar, ValType);
+        auto typeData =
+          getLogicalStorageTypeData(SGF.SGM, getTypeData().AccessKind, ValType);
+
+        // Get the address of the storage property.
+        ManagedValue proj;
+        if (BaseFormalType->mayHaveSuperclass()) {
+          RefElementComponent REC(backingVar, LValueOptions(), varStorageType,
+                                  typeData);
+          proj = std::move(REC).project(SGF, loc, base);
+        } else {
+          assert(BaseFormalType->getStructOrBoundGenericStruct());
+          StructElementComponent SEC(backingVar, varStorageType, typeData);
+          proj = std::move(SEC).project(SGF, loc, base);
+        }
+
+        // Create the allocating initializer function. It captures the metadata.
+        auto delegateInfo = getAttachedPropertyDelegateInfo(field);
+        auto ctor = delegateInfo.initialValueInit;
+        SubstitutionMap subs = backingVar->getType()->getMemberSubstitutionMap(
+                        SGF.getModule().getSwiftModule(), ctor);
+
+        Type ity = ctor->getInterfaceType();
+        AnyFunctionType *substIty = ity.subst(subs)->castTo<AnyFunctionType>();
+
+        auto initRef = SILDeclRef(ctor, SILDeclRef::Kind::Allocator)
+          .asForeign(requiresForeignEntryPoint(ctor));
+        RValue initFuncRV =
+          SGF.emitApplyPropertyDelegateAllocator(loc, subs,initRef,
+                                                 backingVar->getType(),
+                                                 CanAnyFunctionType(substIty));
+        ManagedValue initFn = std::move(initFuncRV).getAsSingleValue(SGF, loc);
+
+        // Create the allocating setter function. It captures the base address.
+        auto setterInfo = SGF.getConstantInfo(setter);
+        SILValue setterFRef = SGF.emitGlobalFunctionRef(loc, setter, setterInfo);
+        auto setterSubs = SGF.getFunction().getForwardingSubstitutionMap();
```

Comment:
Perhaps this will just be the same as `subs` above -- its better to use that instead

Reply:
No, these are different substitutions, because the initializer is from the delegate struct/class, the setter is from the containing struct/class (the setter doesn't have any substitutions, except the forwarding ones from SGF.getFunction()),